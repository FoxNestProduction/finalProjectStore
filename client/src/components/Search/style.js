export const stylesSearch = {
  color: 'primary.main',
  bgcolor: 'background.default',
  width: '100%',
  height: {
    mobile: '51px',
    tablet: '53px',
    desktop: '60px',
  },
};

export const stylesBorder = {
  '&:hover .MuiOutlinedInput-notchedOutline': {
    // borderRadius: 20,
    borderColor: 'primary.hover',
  },
};

export const stylesBtn = {
  color: 'primary.main',
  width: '50%',
  bgcolor: 'background.default',
  '&:hover': {
    bgcolor: 'primary.hover',
    color: 'text.primaryLight',
    borderColor: 'primary.hover',
  },
  '&.Mui-selected': {
    '&:hover': {
      bgcolor: 'primary.hover',
      color: 'text.primaryLight',
    },
    color: 'text.primaryLight',
    bgcolor: 'primary.main',
  },
  padding: '16px 24px',
  border: (theme) => `1px solid ${theme.palette.primary.main}`,
  fontFamily: 'fontPoppins',
  fontSize: {
    mobile: '12px',
    tablet: '14px',
    desktop: '24px',
  },
  height: {
    mobile: '51px',
    tablet: '53px',
    desktop: '60px',
  },
};

export const stylesWrap = {
  p: '10px',
  gap: 2,
};
