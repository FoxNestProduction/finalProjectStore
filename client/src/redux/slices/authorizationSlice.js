import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  isUserAuthorized: false,
  token: null,
};

const authorizationSlice = createSlice({
  name: 'authorization',
  initialState,
  reducers: {
    setAuthorization(state, action) {
      state.isUserAuthorized = action.payload;
    },
    setToken: (state, action) => {
      state.token = action.payload;
    },
  },

});

export const {
  setAuthorization,
  setToken,
} = authorizationSlice.actions;

export default authorizationSlice.reducer;
