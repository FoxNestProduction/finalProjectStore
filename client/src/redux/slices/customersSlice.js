import { createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  customers: [],
};

/* eslint-disable no-param-reassign */

const customersSlice = createSlice({
  name: 'customers',
  initialState,
  reducers: {
    setCustomers(state, action) { // eslint-disable-line no-shadow
      state.customers = action.payload;
    },
  },
});

export const getCustomers = () => async (dispatch) => {
  try {
    const { data } = await axios.get('http://localhost:4000/api/customers');
    console.log(data);
    dispatch(setCustomers(data));// eslint-disable-line no-use-before-define
  } catch (error) {
    console.log('%cError loading customers:', 'color: red; font-weight: bold;', error);
  }
};

export const { setCustomers } = customersSlice.actions;

/* eslint-enable no-param-reassign */

export default customersSlice.reducer;
