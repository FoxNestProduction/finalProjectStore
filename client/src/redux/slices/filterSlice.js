import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { instance } from '../../API/instance';
import { resetSearch, setInputSearchValue, setSearch } from './searchSlice';
import { setError, setLoading } from '../extraReducersHelpers';

export const fetchFilteredProducts = createAsyncThunk(
  'filter/fetchFilteredProducts',
  async (queryString, { rejectWithValue, dispatch }) => {
    try {
      const response = await instance.get(`/products/filter${queryString}`);
      if (response.data.products.length > 0) {
        // dispatch(setSearch([]));
        // dispatch(setInputSearchValue(''));
        dispatch(resetSearch());
      }
      console.log(response.data);
      return response.data;
    } catch (err) {
      return rejectWithValue(err.response.data);
    }
  },
);

const initialState = {
  filteredProducts: [],
  productsQuantity: null,
  loading: false,
  error: null,
  nothingFound: false,
  filterParams: {
    filterCategories: [],
    isTrending: false,
    rating: null, // mostPopular
    isHealthy: false,
    isSupreme: false,
    minPrice: 0,
    maxPrice: 30,
    sort: '',
    startPage: 1,
    perPage: 10,
  },
};

const filterSlice = createSlice({
  name: 'filter',
  initialState,
  reducers: {
    setFilteredProducts(state, action) {
      state.filteredProducts = action.payload;
    },
    setProductsQuantity(state, action) {
      state.productsQuantity = action.payload;
    },
    setNothingFound(state, action) {
      state.nothingFound = action.payload;
    },
    setFilterParams(state, action) {
      console.log('action.payload', action.payload);
      const newState = { ...state.filterParams, ...action.payload };
      console.log('newState', newState);
      state.filterParams = newState;
      // state.filterParams = { ...state.filterParams, ...action.payload };
    },
    resetFilter(state) {
      state.filteredProducts = [];
      state.productsQuantity = null;
      state.nothingFound = false;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchFilteredProducts.pending, setLoading)
      .addCase(fetchFilteredProducts.fulfilled, (state, action) => {
        if (action.payload.length === 0) {
          state.filteredProducts = [];
          state.nothingFound = true;
        } else {
          state.filteredProducts = action.payload.products;
          state.productsQuantity = action.payload.productsQuantity;
          state.nothingFound = false;
        }
        state.loading = false;
      })
      .addCase(fetchFilteredProducts.rejected, setError);
  },
});

export const { setFilteredProducts, setFilterParams,
  setNothingFound, setProductsQuantity, resetFilter } = filterSlice.actions;

export default filterSlice.reducer;
