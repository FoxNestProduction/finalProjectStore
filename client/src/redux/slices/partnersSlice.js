import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { instance } from '../../API/instance';
import { setError, setLoading } from '../extraReducersHelpers';

export const fetchTopPartners = createAsyncThunk(
  'partners/fetchTopPartners',
  async (count, { rejectWithValue }) => {
    try {
      const response = await instance.get(`/partners/filter?perPage=${count}&sort=-rating`);
      return response.data.partners;
    } catch (err) {
      return rejectWithValue(err.response.data);
    }
  },
);

export const getAllPartnersNames = () => async (dispatch) => {
  try {
    const { data } = await instance.get('/partners/names');
    dispatch(setAllPartnersNames(data)); // eslint-disable-line
  } catch (error) {
    console.log('%cError loading allPartnersNames:', 'color: red; font-weight: bold;', error);
  }
};

const initialState = {
  partners: [],
  topPartners: [],
  allPartnersNames: [],
  loading: false,
  error: null,
};

const partnersSlice = createSlice({
  name: 'partners',
  initialState,
  reducers: {
    setPartners(state, action) {
      state.partners = action.payload;
    },
    setAllPartnersNames(state, action) {
      state.allPartnersNames = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchTopPartners.pending, setLoading)
      .addCase(fetchTopPartners.fulfilled, (state, action) => {
        state.loading = false;
        state.topPartners = action.payload;
      })
      .addCase(fetchTopPartners.rejected, setError);
  },
});

export const { setPartners, setAllPartnersNames } = partnersSlice.actions;

// todo: видалити, коли всі дані будуть завантажуватись через asyncThunk/локально в компонентах
// export const getPartners = () => async (dispatch) => {
//   try {
//     const { data } = await instance.get('/partners');
//     dispatch(setPartners(data));
//   } catch (err) {
//     console.log('%cError loading products:', 'color: red; font-weight: bold;', err);
//   }
// };

export default partnersSlice.reducer;
