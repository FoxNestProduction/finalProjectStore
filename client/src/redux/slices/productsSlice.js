import { createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  products: [],
};

/* eslint-disable no-param-reassign */

const productsSlice = createSlice({
  name: 'products',
  initialState,
  reducers: {
    setProducts(state, action) { // eslint-disable-line no-shadow
      state.products = action.payload;
    },
  },
});

export const getProducts = () => async (dispatch) => {
  try {
    const { data } = await axios.get('http://localhost:4000/api/products');
    console.log(data);
    console.log('hello');
    dispatch(setProducts(data));// eslint-disable-line no-use-before-define
  } catch (error) {
    console.log('Error loading products:', error);
  }
};

export const { setProducts } = productsSlice.actions;

/* eslint-enable no-param-reassign */

export default productsSlice.reducer;
