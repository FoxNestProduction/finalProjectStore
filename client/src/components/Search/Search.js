import React, { useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useNavigate } from 'react-router';
import { debounce } from '@mui/material/utils';
import { Autocomplete, InputAdornment, Stack, TextField } from '@mui/material';
import ToggleButton from '@mui/material/ToggleButton';
import ToggleButtonGroup from '@mui/material/ToggleButtonGroup';
import SearchIcon from '@mui/icons-material/Search';
import PropTypes from 'prop-types';
import { stylesSearch, stylesBtn, stylesWrap, stylesBorder } from './style';
import {
  setSearch,
  setKey,
  setInputSearchValue,
  fetchSearchedProductsOrPartners,
  resetSearch,
} from '../../redux/slices/searchSlice';
import { setScrollAnchor } from '../../redux/slices/scrollAnchorSlice';
import { resetFilterParams, deleteFilteredData } from '../../redux/slices/filterSlice';
import { fetchAllProductsNames } from '../../redux/slices/productsSlice';
import { fetchAllPartnersNames } from '../../redux/slices/partnersSlice';

const Search = ({ resetFiltersLocalState }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const inputSearchValue = useSelector((state) => state.search.inputSearchValue);

  useEffect(() => {
    dispatch(fetchAllProductsNames());
    dispatch(fetchAllPartnersNames());
  }, [dispatch]); // eslint-disable-line

  const key = useSelector((state) => state.search.key);
  const allProductsNames = useSelector((state) => state.products.allProductsNames);
  const allPartnersNames = useSelector((state) => state.partners.allPartnersNames);
  const labelForTextField = `Search  ${key}`;

  const handleChangeButton = (event, keyBtn) => {
    if (keyBtn !== null) {
      dispatch(setKey(keyBtn));
      dispatch(setInputSearchValue(''));
      dispatch(setSearch([]));
      dispatch(resetSearch());
    }
  };

  const handleInputChange = async (event, newInputValue) => {
    dispatch(setInputSearchValue(newInputValue));
  };
  const debounceSearch = useRef(
    debounce((inputValue, keyBtn) => {
      if (inputValue.length === 0) {
        dispatch(setSearch([]));
      } else {
        const fetchData = {
          url: keyBtn === 'food' ? '/products/search' : '/partners/search',
          body: {
            query: inputValue,
          },
        };
        try {
          dispatch(fetchSearchedProductsOrPartners(fetchData));
          dispatch(deleteFilteredData());
          dispatch(resetFilterParams());
          resetFiltersLocalState();
          navigate('');
        } catch (err) {
          console.error(`Error getting ${inputValue}: `, err);
        }
      }
    }, 1000),
  );

  useEffect(() => {
    debounceSearch.current(inputSearchValue, key);
  }, [inputSearchValue, key]); // eslint-disable-line

  const topProductsAnchor = useRef();
  useEffect(() => {
    dispatch(setScrollAnchor(topProductsAnchor.current));
  }, [dispatch]);

  return (
    <Stack sx={stylesWrap}>
      <Stack spacing={2} sx={{ stylesSearch }}>
        <Autocomplete
          inputValue={inputSearchValue}
          onInputChange={handleInputChange}
          freeSolo
          id="search"
          disableClearable
          options={key === 'food' ? allProductsNames : allPartnersNames}
          renderInput={(params) => (
            <TextField
              sx={stylesBorder}
              {...params}
              label={labelForTextField}
              variant="outlined"
              InputProps={{
                ...params.InputProps,
                type: 'search',
                startAdornment: (
                  <InputAdornment position="start">
                    <SearchIcon />
                  </InputAdornment>
                ),
              }}
            />
          )}
        />
      </Stack>

      <ToggleButtonGroup value={key} exclusive onChange={handleChangeButton} aria-label="Platform" ref={topProductsAnchor}>
        <ToggleButton value="food" sx={stylesBtn}>
          Food
        </ToggleButton>
        <ToggleButton value="restaurant" sx={stylesBtn}>
          Restaurant
        </ToggleButton>
      </ToggleButtonGroup>
    </Stack>
  );
};

Search.propTypes = {
  resetFiltersLocalState: PropTypes.func,
};

Search.defaultProps = {
  resetFiltersLocalState: () => {},
};

export default Search;
