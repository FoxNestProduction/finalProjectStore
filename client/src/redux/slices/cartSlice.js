import { createSlice } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import axios from 'axios';
import { allProducts } from './productsSlice';

const initialState = {
  cartItems: [],
  isLoading: false,
  isCart: false,
};

/* eslint-disable no-param-reassign */

const cartSlice = createSlice({
  name: 'cart',
  initialState,
  reducers: {
    addToCart(state, action) {
      // 1-ий варіант - коли нам в action.payload надходить повний об'єкт item
      if (state.cartItems.length === 0 && action.payload !== null) {
        state.cartItems.push(action.payload);
      } else {
        const index = state.cartItems.findIndex((item) => item.id === action.payload.id);
        if (index === -1) {
          state.cartItems.push(action.payload);
        } else {
          state.cartItems[index].quantity += 1;
        }
      }
      /** 2-ий варіант, коли нам приходить суто id товара потрібно звернутись
       * до однієї з двох написаних нижче функцій:
       * - deleteOrAddCartByItemId , або
       * - deleteOrAddFromCartByItemIdWithValueFromState
       * Потрібно буде протестувати, який варіант зручніше, той і використовувати
       */
    },
    setCart(state, action) {
      if (state.cartItems.length === 0) {
        state.cartItems = action.payload;
      } else {
        const uniqueFilteredProducts = action.payload.products.filter((product) => {
          const matchedProduct = state.cartItems
            .find((cartItem) => cartItem.item.id !== product.id);
          const mark = matchedProduct !== undefined;
          return mark;
        });
        const notUniqueFilteredProducts = action.payload.products.filter((product) => {
          const matchedProduct = state.cartItems
            .find((cartItem) => cartItem.item.id === product.id);
          const mark = matchedProduct !== undefined;
          return mark;
        }).map((product) => {
          const matchedProduct = state.cartItems
            .find((cartItem) => cartItem.item.id === product.id);
          product.quantity += matchedProduct.quantity;
          return product;
        });
        state.cartItems = { ...uniqueFilteredProducts, ...notUniqueFilteredProducts };
      }
    },
    setIsLoading(state, action) {
      state.isLoading = action.payload;
    },
    setIsCart(state, action) {
      state.isCart = action.payload;
    },
    deleteFromCart(state, action) {
      // 1 - ий варіант - коли нам в action.payload надходить повний об'єкт item
      if (state.cartItems.length) {
        const index = state.cartItems.findIndex((item) => item.id === action.payload.id);

        if (index !== -1) {
          state.cartItems[index] = (
            state.cartItems[index] === 1
              ? state.cartItems.splice(index, 1)
              : state.cartItems[index].quantity -= 1
          );
        }
      }
      /** 2-ий варіант, коли нам приходить суто id товара потрібно звернутись
         * до однієї з двох написаних нижче функцій:
         * - deleteOrAddCartByItemId , або
         * - deleteOrAddFromCartByItemIdWithValueFromState
         * Потрібно буде протестувати, який варіант зручніше, той і використовувати
         */
    },
  },
});

export const { addToCart, setCart, setIsLoading, deleteFromCart, setIsCart } = cartSlice.actions;

export const getCartItemsFromServer = () => async (dispatch) => {
  try {
    dispatch(setIsLoading(true));
    const userId = useSelector((state) => state.user.user.id);
    const { data } = await axios.get('http://localhost:4000/cart');

    if (data !== null && data !== undefined) {
      const cartContainer = data.find((cartObj) => cartObj.customerId.id === userId);

      if (cartContainer !== null && data !== undefined) {
        dispatch(setIsCart(true));
        dispatch(setCart(cartContainer));
      } else {
        dispatch(setIsCart(false));
      }
    } else {
      dispatch(setIsCart(false));
    }
    // якщо data undefined, то нічого не передаємо в кошик
    // а isCart(true or false). Код далі не виконується і дані на сторінці відмальовуються
    // тільки зі стора

    // dispatch(setCart(cartContainer)); // якщо ми тут оновимо state то оновиться і localStorage
    //  тому тут потрібно написати логіку злиття даних з localStorage з даними з сервера.
    dispatch(setIsLoading(false));
  } catch (error) {
    console.warn('Error loading cart:', error);
    dispatch(setIsLoading(false));
    dispatch(setIsCart(false));
  }
};

/** 2 варіанти функції якщо ми додаємо або видаляємо товар в state cart через прокидування id:
 * TODO: варіант A:
 * ми робимо підписку на необхідний нам state всіх продуктів в самій функції.
 * Знаходимо по id необхідний нам об'єкт, та прокидуємо його в один з reducer-ов
 * за допомогою одного з ключів DELETE або ADD. Необхідний нам об'єкт падає
 * в визначений reducer та додається до кошику, або ж видаляється з нього.
 * Ця функція передається через dispatch. В неї прокидується id та key.
 * dispatch(deleteOrAddCartByItemId(id, key))
 */
export const deleteOrAddCartByItemId = (id, key) => (dispatch) => {
  const products = useSelector((state) => state.products.products);
  if (products.length !== 0) {
    const cartItem = products.find((product) => product.id === id);
    if (cartItem !== null && cartItem !== undefined) {
      if (key === 'ADD') {
        dispatch(addToCart(cartItem));
      }
      if (key === 'DELETE') {
        dispatch(deleteFromCart(cartItem));
      }
    }
  }
};

/** 2 варіанти функції якщо ми додаємо або видаляємо товар в state cart через прокидування id:
 * TODO: варіант B:
 * всередині компонента товара ми робимо підписку через useSelector на цей самий об'єкт товара.
 * Якщо точніше, то це не зовсім підписка, а пошук через функціональний селектор необхідних нам
 * даних в останній версії store, завдяки функції useSelector. В самому компоненті потрібно буде
 * прописати такий код:
 * TODO: const fullItemObject = useSelector(deleteOrAddFromCartByItemIdWithValueFromState(id))
 * в константу fullItemObject прийде повний об'єкт цього товара, і при кліку на кнопку
 * "Додати до кошика" (в нашому випадку це знак +), або видалення з кошика
 * ми передаємо в onClick такий код:
 * додати до кошика
 * TODO: onClick={dispatch(addToCart(fullItemObject))}
 * або такий, якщо видалити з кошика
 * TODO: onClick={dispatch(deleteFromCart(fullItemObject))}
 * Попередньо імпортувавши dispatch та addToCart()/deleteFromCart()
 */

export const deleteOrAddFromCartByItemIdWithValueFromState = (id) => (state) => {
  const products = allProducts(state);
  console.log(products);
  if (products.length !== 0) {
    const cartItem = products.find((product) => product.id === id);
    console.log(cartItem);
    if (cartItem !== null && cartItem !== undefined) {
      return cartItem;
    }
    return null;
  }
  return null;
};

export default cartSlice.reducer;
