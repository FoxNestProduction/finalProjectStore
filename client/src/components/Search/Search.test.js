import React from 'react';
import { fireEvent, render, waitFor, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router, useNavigate } from 'react-router-dom';
import configureStore from 'redux-mock-store';

import Search from './Search';
import { fetchAllProductsNames } from '../../redux/slices/productsSlice';
import { fetchAllPartnersNames } from '../../redux/slices/partnersSlice';
import { deleteFilteredData, resetFilterParams } from '../../redux/slices/filterSlice';
import {  
  setKey,
  setInputSearchValue,
  setSearch,
  resetSearch,
  fetchSearchedProductsOrPartners,
} from '../../redux/slices/searchSlice';

const mockStore = configureStore();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../../redux/slices/productsSlice', () => ({
  ...jest.requireActual('../../redux/slices/productsSlice'),
  fetchAllProductsNames: jest.fn(),
}));

jest.mock('../../redux/slices/partnersSlice', () => ({
  ...jest.requireActual('../../redux/slices/partnersSlice'),
  fetchAllPartnersNames: jest.fn(),
}));

jest.mock('../../redux/slices/searchSlice', () => ({
  ...jest.requireActual('../../redux/slices/searchSlice'),
  setKey: jest.fn(),
  setInputSearchValue: jest.fn(),
  setSearch: jest.fn(),
  resetSearch: jest.fn(),
  fetchSearchedProductsOrPartners: jest.fn(),
}));

jest.mock('../../redux/slices/filterSlice', () => ({
  ...jest.requireActual('../../redux/slices/filterSlice'),
  deleteFilteredData: jest.fn(),
  resetFilterParams: jest.fn(),
}));

const mockDispatch = jest.spyOn(require('react-redux'), 'useDispatch');

describe('Snapshot test', () => {
  const store = mockStore({
    search: {
      search: [],
      key: 'food',
      inputSearchValue: '',
    },
    partners: {
      allPartnersNames: ['partner1', 'partner2'],
    },
    products: {
      allProductsNames: ['product1', 'product2'],
    },
  });

  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(false);
  });

  test('should Search render', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    const navigateMock = jest.fn();
    useNavigate.mockReturnValue(navigateMock);

    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <Search />
        </Router>
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Food')).toBeInTheDocument();
    expect(screen.getByText('Restaurant')).toBeInTheDocument();
  });

  test('should update input value on user input', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    const navigateMock = jest.fn();
    useNavigate.mockReturnValue(navigateMock);

    render(
      <Provider store={store}>
        <Router>
          <Search />
        </Router>
      </Provider>,
    );

    expect(dispatch).toHaveBeenCalledWith(setInputSearchValue('new search value'));
  });

  test('should dispatch actions on button toggle', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    const navigateMock = jest.fn();
    useNavigate.mockReturnValue(navigateMock);

    render(
      <Provider store={store}>
        <Router>
          <Search />
        </Router>
      </Provider>,
    );

    const foodButton = screen.getByText('Food');
    fireEvent.click(foodButton);

    expect(dispatch).toHaveBeenCalledWith(setKey('food'));
    expect(dispatch).toHaveBeenCalledWith(setInputSearchValue(''));
    expect(dispatch).toHaveBeenCalledWith(setSearch([]));
    expect(dispatch).toHaveBeenCalledWith(resetSearch());
  });

  test('should fetch data on input change', async () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    const navigateMock = jest.fn();
    useNavigate.mockReturnValue(navigateMock);

    render(
      <Provider store={store}>
        <Router>
          <Search />
        </Router>
      </Provider>,
    );

    await waitFor(() => {
      expect(dispatch).toHaveBeenCalledWith(fetchSearchedProductsOrPartners(expect.any(Object)));
    });
  });

  test('should update Redux state after actions are dispatched', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    const navigateMock = jest.fn();
    useNavigate.mockReturnValue(navigateMock);

    render(
      <Provider store={store}>
        <Router>
          <Search />
        </Router>
      </Provider>,
    );

    const foodButton = screen.getByText('Food');
    fireEvent.click(foodButton);

    const updatedState = store.getState();
    expect(updatedState.search.key).toEqual('food');
    expect(updatedState.search.inputSearchValue).toEqual('');
  });
});
