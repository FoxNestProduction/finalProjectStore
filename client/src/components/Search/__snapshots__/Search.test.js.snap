// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test should Search render 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-z7kfs3-MuiStack-root"
  >
    <div
      class="MuiStack-root css-zpwguh-MuiStack-root"
    >
      <div
        class="MuiAutocomplete-root css-1h51icj-MuiAutocomplete-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-kifgi1-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="search"
            id="search-label"
          >
            Search  food
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiAutocomplete-inputRoot css-8j6b76-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </div>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch MuiInputBase-inputAdornedStart MuiAutocomplete-input MuiAutocomplete-inputFocused css-aeeyth-MuiInputBase-input-MuiOutlinedInput-input"
              id="search"
              role="combobox"
              spellcheck="false"
              type="search"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Search  food
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-label="Platform"
      class="MuiToggleButtonGroup-root css-1gjgmky-MuiToggleButtonGroup-root"
      role="group"
    >
      <button
        aria-pressed="true"
        class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-1ltdkv-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="food"
      >
        Food
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-1ltdkv-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="restaurant"
      >
        Restaurant
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
